### PR TITLE
support now() for sharding value

### PR DESCRIPTION
**Pull request by Pramy**
Change-Id: I3fb454efc9089e10f1e3a8c5d4eb575327816548

support #1497

### CODE REVIEW

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/DefaultTimeService.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        }
+        return Optional.of(new Date());
+    }
+
+    /**
+     * judge expression.
+     * @param segment ExpressionSegment
+     * @return true or false
+     */
+    public boolean isNowExpression(final ExpressionSegment segment) {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/DefaultTimeService.java]: Can you decouple ExpressionSegment and DefaultTimeService?
DefaultTimeService should keep same method signature with TimeService.
The logic of aware ExpressionSegment is now or not should be in parse module.

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/TimeService.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+
+import java.util.Date;
+
+public interface TimeService {
+
+    /**
+     * get date from time server.
+     * @return date
+     */
+    Optional<Date> getTime();
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/TimeService.java]: Why getTime return Optional? Which branch will return Optional date?

File: sharding-core/sharding-core-route/src/test/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueBetweenOperatorGeneratorTest.java [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+
+import java.util.Calendar;
+import java.util.Date;
+import java.util.LinkedList;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+import static org.junit.Assert.assertFalse;
+
+public class ConditionValueBetweenOperatorGeneratorTest {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/test/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueBetweenOperatorGeneratorTest.java]: Should add final if class doesn't design for extend.

File: sharding-core/sharding-core-route/src/test/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueInOperatorGeneratorTest.java [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        segmentCollection.add(new CommonExpressionSegment(0, 0, "now()"));
+        PredicateInRightValue inRightValue = new PredicateInRightValue(segmentCollection);
+        Optional<RouteValue> optional = generator.generate(inRightValue, column, new LinkedList<>());
+        assertTrue(optional.isPresent());
+        assertTrue(optional.get() instanceof ListRouteValue);
+        ListRouteValue listRouteValue = (ListRouteValue) optional.get();
+        assertFalse(listRouteValue.getValues().isEmpty());
+        assertTrue(listRouteValue.getValues().iterator().next() instanceof Date);
+    }
+
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/test/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueInOperatorGeneratorTest.java]: Should remove useless blank line.

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/DefaultTimeService.java (hunk 1 above)
Comment (Pramy) [reply sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/DefaultTimeService.java]: thanks for your suggestion. This method is used in many places. Maybe I can create a `NowExpressionUtils` class?

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/TimeService.java (hunk 2 above)
Comment (Pramy) [reply sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/TimeService.java]: I have changed it to `Date`

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/ExpressionConditionUtils.java [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.core.route.router.sharding.condition;
+
+import org.apache.shardingsphere.core.parse.sql.segment.dml.expr.ExpressionSegment;
+import org.apache.shardingsphere.core.parse.sql.segment.dml.expr.complex.ComplexExpressionSegment;
+
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/ExpressionConditionUtils.java]: Missing javadoc

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/ExpressionConditionUtils.java [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+public final class ExpressionConditionUtils {
+
+    /**
+     * judge now() expression.
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/ExpressionConditionUtils.java]: First word of javadoc should be uppercase.

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueBetweenOperatorGenerator.java [hunk 7]
```diff
@@ -39,7 +42,15 @@
     public Optional<RouteValue> generate(final PredicateBetweenRightValue predicateRightValue, final Column column, final List<Object> parameters) {
         Optional<Comparable> betweenRouteValue = new ConditionValue(predicateRightValue.getBetweenExpression(), parameters).getValue();
         Optional<Comparable> andRouteValue = new ConditionValue(predicateRightValue.getAndExpression(), parameters).getValue();
-        return betweenRouteValue.isPresent() && andRouteValue.isPresent()
-                ? Optional.<RouteValue>of(new RangeRouteValue<>(column.getName(), column.getTableName(), Range.closed(betweenRouteValue.get(), andRouteValue.get()))) : Optional.<RouteValue>absent();
+        if (betweenRouteValue.isPresent() && andRouteValue.isPresent()) {
+            return Optional.<RouteValue>of(new RangeRouteValue<>(column.getName(), column.getTableName(), Range.closed(betweenRouteValue.get(), andRouteValue.get())));
+        }
+        DefaultTimeService defaultTimeService = new DefaultTimeService();
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueBetweenOperatorGenerator.java]: Do we need SPI to load `TimeService`?

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/DefaultTimeService.java [hunk 8]
```diff
@@ -0,0 +1,43 @@
...
+ */
+
+package org.apache.shardingsphere.core.route.spi;
+
+import org.apache.shardingsphere.core.spi.NewInstanceServiceLoader;
+
+import java.util.Collection;
+import java.util.Date;
+
+public final class DefaultTimeService implements TimeService {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/DefaultTimeService.java]: Please make `DefaultTimeService` as a simple SPI implementation and let caller to call SPI

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/TimeService.java [hunk 9]
```diff
@@ -0,0 +1,29 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.shardingsphere.core.route.spi;
+
+import java.util.Date;
+
+public interface TimeService {
```
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/TimeService.java]: Missing javadoc

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/DefaultTimeService.java (hunk 8 above)
Comment (terrymanu) [inline sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/spi/DefaultTimeService.java]: Missing javadoc

File: sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueBetweenOperatorGenerator.java (hunk 7 above)
Comment (Pramy) [reply sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueBetweenOperatorGenerator.java]:
```
sharding clomn is create_time
select * from ts_order where create_time between ? and now()
```
I think if use this to query , we still need SPI to load `TimeService`

Comment (terrymanu) [reply sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueBetweenOperatorGenerator.java]: What I mean is we need SPI to load `TimeService`, but we should load SPI obvious from the caller.
Right now, we put loading SPI logic into `DefaultTimeService`, that make `DefaultTimeService` to complicated. 
`DefaultTimeService` may just as a regular SPI implementation.

Comment (Pramy) [reply sharding-core/sharding-core-route/src/main/java/org/apache/shardingsphere/core/route/router/sharding/condition/generator/impl/ConditionValueBetweenOperatorGenerator.java]: I get what you meant now

**Pull request merged**